_drafts/services-commands-y-otros-poderosos-patrones-en-rails.md

```diff
@@ -0,0 +1,133 @@
+
+## Services
+
+El concepto de Services en Rails goza de cierta difusión en la comunidad, apareciendo junto con Rails 4.  Como lo define [Dave Copeland en su artículo dedicado](http://multithreaded.stitchfix.com/blog/2015/06/02/anatomy-of-service-objects-in-rails/), el rol de un service es **tener el código de un porción especifica de la lógica de negocio**.  El service es una clase que expone uno o varios métodos públicos para llevar a cabo tareas puntuales. Si son más de uno, se enmarcan dentro de un contexto común que describe el nombre del servicio.  Este nombre debe ser un sustantivo, seguido de “_service”.  
+
+
+Imaginemos una aplicación para escuchar música via streaming, y que recomienda al usuario canciones y artistas según las canciones que ha escuchado.  En este contexto, creamos un servicio para obtener las canciones o artistas relacionados, en base a una canción dada.
+
+```ruby
+class RelatedMediaService
+
+def initialize source
+ @source = source
+end
+
+def get_songs
+ related_songs  = []
+ related_songs << Song.where(genre: @source.genre)
+ related_songs << Song.where(year: @source.year, country: @source.country)
+ #some other song relating bussiness logic
+ related_songs
+end
+
+def get_artists
+ related_artist = []
+ related_artist << Artist.where(genre: @source.genre)
+ #some other artist relating bussiness logic
+
+end  
+```
+Aquí vemos que la clase opera como una función:  recibe un input, ejecuta una operación, y da un output.  En cuanto al output, podemos optar entre que sea a través del retorno del método perform, o bien no retornar nada si no que el servicio internamente modifique los objetos sobre los que actúa.
+
+Para trabajar con estos poderosos tipos de  clases, en Platanus desarrollamos la gema [Power Types](https://github.com/platanus/power-types).  Esta gema nos crea las carpetas y clases base para crear Services con una estructura definida.
+
+## Commands
+Los commands también están a nuestra disposición.  Corresponden a una versión más funcional de services.  Se usan para operaciones acotadas y no excesivamente complejas.  Un command solo expone un método perform, que ejecuta la acción que describe el nombre del command.  Este nombre, a diferencia de los services, debe ser un verbo. Y también se diferencia ya que no debe guardar estado si no que ejecutarse de una vez a partir de los argumentos que recibe al inicializarse.
+
+Los services y commands atacan objetivos similares, tal vez todavía son difusas para ti las diferencias.  Nos podemos hacer las siguientes preguntas para identificar cuando corresponde usar un service en vez de un command:
+
+* ¿La clase que vamos a crear debe ofrecer más de una función, y por lo tanto exponer 2 o más métodos públicos?
+
+* ¿Necesitamos una clase que guarde su estado, y se hacen varias llamadas en distintos momentos a una misma instancia de ella?
+
+* ¿La lógica de la implementación es excesivamente compleja, tiene llamados a APIs o a alguna librería sofisticada?
```

Comment:
Buena, gracias..
@felipe, comentame si vaz a hacer algun cambio en esas preguntas y lo
mergeamos...

On Fri, Aug 12, 2016 at 12:33 PM Ignacio Baixas <notifications@github.com>
wrote:

> In _drafts/services-commands-y-otros-poderosos-patrones-en-rails.md
> <https://github.com/platanus/blog/pull/68#discussion_r74620801>:
>
> > +end
> > +```
> > +Aquí vemos que la clase opera como una función:  recibe un input, ejecuta una operación, y da un output.  En cuanto al output, podemos optar entre que sea a través del retorno del método perform, o bien no retornar nada si no que el servicio internamente modifique los objetos sobre los que actúa.
> > +
> > +Para trabajar con estos poderosos tipos de  clases, en Platanus desarrollamos la gema [Power Types](https://github.com/platanus/power-types).  Esta gema nos crea las carpetas y clases base para crear Services con una estructura definida.
> > +
> > +## Commands
> > +Los commands también están a nuestra disposición.  Corresponden a una versión más funcional de services.  Se usan para operaciones acotadas y no excesivamente complejas.  Un command solo expone un método perform, que ejecuta la acción que describe el nombre del command.  Este nombre, a diferencia de los services, debe ser un verbo. Y también se diferencia ya que no debe guardar estado si no que ejecutarse de una vez a partir de los argumentos que recibe al inicializarse.
> > +
> > +Los services y commands atacan objetivos similares, tal vez todavía son difusas para ti las diferencias.  Nos podemos hacer las siguientes preguntas para identificar cuando corresponde usar un service en vez de un command:
> > +
> > +* ¿La clase que vamos a crear debe ofrecer más de una función, y por lo tanto exponer 2 o más métodos públicos?
> > +
> > +* ¿Necesitamos una clase que guarde su estado, y se hacen varias llamadas en distintos momentos a una misma instancia de ella?
> > +
> > +* ¿La lógica de la implementación es excesivamente compleja, tiene llamados a APIs o a alguna librería sofisticada?
>
> Podría, no se me ocurre la regla exacta para separar un servicio en varios
> commands. Hay dos casos que se me ocurren para mantener todo en un
> servicio:
>
>    1. Hay mucha lógica común entre las funciones provistas.
>    2. Son funciones simples. Si alguna de las funciones tomara mucha
>    complejidad estaría inclinado a meter la lógica en un command y dejar la
>    función en el servicio como wrapper del command.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/platanus/blog/pull/68/files/c994f29df0b6a4af85afac293f58946db7b9718a#r74620801>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAN6xZlsh2Eb9FsBxNUVt2SvWLo8gavUks5qfKBRgaJpZM4JJEPg>
> .
>
-- 

Juan Ignacio Donoso
Co-fundador | Desarrollador | Diseñador
juan.ignacio@platan.us

Platanus: (+562) 2570 9846 <%28%2B562%29%2028818025>
Las Catalpas 1588, Vitacura, Santiago, Chile